datafusion/expr-common/src/signature.rs

```diff
@@ -455,6 +461,46 @@ fn get_data_types(native_type: &NativeType) -> Vec<DataType> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct FunctionSignature {
+    pub parameters: Vec<ParameterSignature>,
+    /// The volatility of the function. See [Volatility] for more information.
+    pub volatility: Volatility,
+}
+
+pub type ParameterSignature = Vec<ParameterType>;
+
+#[derive(Debug, Clone, Eq, PartialOrd, Hash)]
+pub struct ParameterType {
+    pub param_type: LogicalTypeRef,
```

Comment:
`param_type`: target type of function signature
`allowed_casts`: implicit coercion allowed to cast to target type.

For example,
`param_type`: string
`allowed_casts`: binary, int

```
Valid: All are casted to string
func(string) 
func(binary)
func(int)
Invalid:
func(float or other types)
```

Reply:
I would like to have a breaking change to `Signature::Coercible` too, the only concern is whether this cause regression or large impact to downstream

Reply:
@shehabgamin If we replace `CoercibleV2` with `Signature::Coercible` would it be a large change we should be concerned to?

Reply:
If not, I'll replace Signature::Coercible with Signature::CoercibleV2.